src/transport/MessageEndpoint.cpp

```diff
@@ -0,0 +1,113 @@
+#include <faabric/transport/MessageEndpoint.h>
+#include <faabric/util/locks.h>
+
+namespace faabric::transport {
+MessageEndpoint::MessageEndpoint(const std::string& hostIn, int portIn)
+  : host(hostIn)
+  , port(portIn)
+{}
+
+MessageEndpoint::~MessageEndpoint()
+{
+    if (this->socket) {
+        this->close();
+    }
+}
+
+void MessageEndpoint::open(faabric::transport::MessageContext& context,
+                           faabric::transport::SocketType sockType,
+                           bool bind)
+{
+    std::string address =
+      "tcp://" + this->host + ":" + std::to_string(this->port);
+
+    // Note - only one socket may bind, but several can connect. This allows
+    // for easy N - 1 or 1 - N PUSH/PULL patterns. Order between bind and
+    // connect does not matter.
+    switch (sockType) {
+        case faabric::transport::SocketType::PUSH:
+            this->socket = std::make_unique<zmq::socket_t>(
+              context.get(), zmq::socket_type::push);
+            break;
+        case faabric::transport::SocketType::PULL:
+            this->socket = std::make_unique<zmq::socket_t>(
+              context.get(), zmq::socket_type::pull);
+            break;
+        default:
+            throw std::runtime_error("Unrecognized socket type");
+    }
+
+    // Set linger period to zero. This avoids blocking if closing a socket with
+    // outstanding messages.
```

Comment:
Is this a good thing? Shouldn't we block when closing to wait for the messages to be processed?

Reply:
Yes, I think you are right :+1: